src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets

```diff
@@ -158,65 +158,74 @@
   <!-- Generate BuildVersion.props -->
   <!-- #################################### -->
   <Target Name="CreateOrUpdateCurrentVersionFile" Condition="'$(SkipVersionGeneration)'!='true'" DependsOnTargets="GenerateCurrentVersion">
-    <Error Condition="'$(OverrideBuildNumberMajor)'==''" Text="Error when generating current version." />
+    <!-- Error out if GeneratedBuildNumberMajor is empty since GenerateCurrentVersion should have calculated it already. -->
+    <Error Condition="'$(GeneratedBuildNumberMajor)'==''" Text="Error when generating current version." />
 
     <ItemGroup>
       <CurrentVersionLines Include="%3C%3Fxml version=%221.0%22 encoding=%22utf-8%22%3F%3E" />
-      <CurrentVersionLines Include="%3C!-- This is a generated file. To re-generate it, run build.cmd --%3E" />
+      <CurrentVersionLines Include="%3C!-- This is a generated file. --%3E" />
       <CurrentVersionLines Include="%3CProject xmlns=%22http://schemas.microsoft.com/developer/msbuild/2003%22%3E" />
       <CurrentVersionLines Include="%3CPropertyGroup%3E" />
-      <CurrentVersionLines Include="%3CBuildNumberMajor Condition=%22%27%24(BuildNumberMajor)%27==%27%27%22%3E$(OverrideBuildNumberMajor)%3C/BuildNumberMajor%3E" />
+      <CurrentVersionLines Include="%3CBuildNumberMajor Condition=%22%27%24(BuildNumberMajor)%27==%27%27%22%3E$(GeneratedBuildNumberMajor)%3C/BuildNumberMajor%3E" />
       <CurrentVersionLines Include="%3C/PropertyGroup%3E" />
       <CurrentVersionLines Include="%3C/Project%3E" />
       <BuildVersionFileItem Include="$(BuildVersionFile)" />
     </ItemGroup>
-
+    
     <PropertyGroup>
-      <BuildVersionSemaphoreFile>%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)semaphore.txt</BuildVersionSemaphoreFile>
-      <ShouldLockFileWrite Condition="!Exists('$(BuildVersionSemaphoreFile)')">true</ShouldLockFileWrite>
+      <BuildVersionFilePath>%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)</BuildVersionFilePath>
```

Comment:
I would define this in the BuildVersion.txt and use it there as well. No need to use metadata traversal for this.